op-supernode/supernode/activity/heartbeat/heartbeat.go

```diff
@@ -0,0 +1,69 @@
+package heartbeat
+
+import (
+	"context"
+	"crypto/rand"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	gethlog "github.com/ethereum/go-ethereum/log"
+)
+
+// Activity that emits periodic heartbeats and exposes a simple liveness RPC.
+type Heartbeat struct {
+	log      gethlog.Logger
+	interval time.Duration
+	ctx      context.Context
+	cancel   context.CancelFunc
+}
+
+// New creates a new Heartbeat activity.
+func New(log gethlog.Logger, interval time.Duration) *Heartbeat {
+	return &Heartbeat{log: log, interval: interval}
+}
+
+// Start begins the periodic logging loop.
+func (h *Heartbeat) Start(ctx context.Context) error {
+	if h.interval <= 0 {
+		h.interval = time.Second
+	}
+	h.ctx, h.cancel = context.WithCancel(ctx)
+	ticker := time.NewTicker(h.interval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-h.ctx.Done():
+			return h.ctx.Err()
+		case <-ticker.C:
+			h.log.Info("heartbeat")
+		}
+	}
+}
+
+// Stop stops the heartbeat loop.
+func (h *Heartbeat) Stop(ctx context.Context) error {
+	if h.cancel != nil {
+		h.cancel()
+	}
+	return nil
+}
+
+// RPCNamespace returns the JSON-RPC namespace for this activity.
+func (h *Heartbeat) RPCNamespace() string { return "heartbeat" }
+
+// RPCService returns the service object whose exported methods are exposed in RPC.
+func (h *Heartbeat) RPCService() interface{} { return (*api)(h) }
+
+// api hosts JSON-RPC methods for the Heartbeat activity.
+type api Heartbeat
+
+// Check returns a 4-byte random UUID (hex-encoded) for liveness.
```

Comment:
(nitpick) Technically 4 byte random is not a uuid.

Reply:
Ah you are correct because a UUID is defined as having 128 bits. I'll correct the reference here.

Where UUID stands for *Universally* Unique Identifier, I should just call this UID, Unique Identifier.

(*technically* not even 128bit should be able to call itself "Universally Unique". UIDs of any length avoid hash collisions, and while the UUID is statistically guaranteed uniqueness, a 4-byte version is suitable for tracking liveness, as the chance of any 4-byte UIDs colliding in *adjacent messages* is also vanishingly small.